### PR TITLE
fatfs: Clarify the MS-DOS 7.x boot messages

### LINKED ISSUE #681: [INFO] MS-DOS 7.0 Boot message string
I stumbled across this little bit of code whilst poking about in @ecm-pushbx ldebug which seems to explain the meaning of the four bytes preceding the MS-DOS 7.0 boot block message string [msg.asm](https://bitbucket.org/ecm/ldebug/src/6f1572b48e81e8b33525c15e50f8e33ae17142c7/source/msg.asm?fileviewer=file-view-default#msg.asm-908). 
Here's the place in fatfs.c where it's used by Dosemu https://github.com/stsp/dosemu2/blob/0e1cff8a8f71cf5cdd4ce008795b7ba706c8c6b9/src/base/misc/fatfs.c#L1865-L1880

So it seems like there may be other messages to be added and the correct offsets calculated.

### PR DESCRIPTION
**Pull request by andrewbird**
1/ Add further comments to explain the format better.
2/ Add single padding byte to avoid message truncation.
3/ Add comment regarding value for message table location value.

[fixes #681]

### CODE REVIEW

**Comment by andrewbird** [comment]
Yes you are correct that the message is wrong, but I wonder if it's possible to exit dosemu in some way, endlessly trying to boot without changing anything seems pointless?

**Comment by ecm-pushbx** [comment]
You'd have to figure out some heuristic as to what interrupt 19h call is pointless and thus should quit dosemu, as opposed to what interrupt 19h call should actually reboot.

**Comment by stsp** [comment]
> You'd have to figure out some heuristic as to what interrupt 19h call is pointless
> and thus should quit dosemu, as opposed to what interrupt 19h call should actually reboot.

Interesting point.
Can we look at from where the int19 was called?
For example, if it was called from segment 0x70,
then exit?

**Comment by stsp** [comment]
I also wonder if we can probably separate
our "return to linux" msg from those 4 msgs
for DOS?

**Comment by stsp** [comment]
I mean we better add those 4 msgs
only for DOS-7, while keeping our own
msg untouched?

**Comment by ecm-pushbx** [comment]
> For example, if it was called from segment 0x70, then exit?

I think I observed that the early loader relocates itself to the top of low memory first. I can check tomorrow.

**Comment by andrewbird** [comment]
> I also wonder if we can probably separate
our "return to linux" msg from those 4 msgs
for DOS?

That string is specific to the boot loader, it might look generic but it's not used anywhere else. The text could be made specific to MS-DOS 7. It could also only be added if f->sys_type == NEWMSD_D to save space in the boot block, but there is already plenty of space in our generated boot block.

**Comment by stsp** [comment]
> That string is specific to build_boot_blk(), it might look generic but it's not used anywhere else.

Ah, its since you removed `fatfs_boot.S` I suppose,
where it was used in the past. Now we have its equivalent
spread across multiple places (`boot()`, `mimic_boot_blk()`),
and they all provide their own msgs.

> It could also only be added if f->sys_type == NEWMSD_D

I would appreciate if you do that, just to underline
the fact that it is not used elsewhere in the boot process,
which was not always so. I suppose no other DOS re-uses
msgs from the boot sector? Or if some do, do they need
this offset table too?

**Comment by stsp** [comment]
> It could also only be added if f->sys_type == NEWMSD_D

Or, for example, add message always, but
the offset table only for ` NEWMSD_D`?
With the comment that it is otherwise unused?

**Comment by ecm-pushbx** [comment]
> I suppose no other DOS re-uses msgs from the boot sector? Or if some do, do they need this offset table too?

None that I know of, no.

**Comment by stsp** [comment]
> None that I know of, no.

But I suppose most boot sectors contain
that msg for their own use, right? In this
case does the 0x1ee pointer points to an
msg itself or to the offset table?

**Comment by ecm-pushbx** [comment]
Neither, MS-DOS 7 and 8 are the only ones I know of that re-use messages at all. So that pointer also doesn't exist otherwise.

**Comment by stsp** [comment]
> Neither

But neither is impossible, I suppose every
boot sector contains the failure msg somewhere.
It may be just that there is no pointer to it in any
known location, and it is hardcoded somewhere
in the boot code.

**Comment by ecm-pushbx** [comment]
> It may be just that there is no pointer to it in any known location, and it is hardcoded somewhere in the boot code.

Yes, that seems to be the case.

For reference, ldosboot makes do with single-letter error messages because of the space pressure.

**Comment by stsp** [comment]
OK, so in this case I think the best
solution would be to add the msg only
for `NEWMSD_D`, and add the comment
for the rest that it can be added here if
need be. In the past, we had it used in
`fatfs_boot.S`.

**Comment by stsp** [comment]
> Just looking at read_boot() I wonder why in commit dd149c9 @stsp
> you moved dosemu boot block specific code out of build_boot_blk() and put it there instead?

I don't know exactly why I did that (was too
long ago), but at the end of the day, `build_boot_blk()`
is called from `read_boot()`, so the change is
effectively a no-op. But looking into this now, I
see you saying "dosemu boot block specific code",
which it is actually not: it was a hack you added
to pass NDD test. I think my idea was probably
that `build_boot_blk()` only builds the boot code
block, and the rest of `read_boot()` builds the
data block. This hack belongs to the data blocks,
at least its not executed. In any case, if it hurts you,
I think no problems to get it back.
Note that when I did so, I've also filled #524 with
the idea how I see the boot block handling in the
future. dd149c9 was probably a (very small) part
of that idea, or maybe even not.

File: src/base/misc/fatfs.c [hunk 1]
```diff
@@ -1895,7 +1889,15 @@ void build_boot_blk(fatfs_t *f, unsigned char *b)
      illegal displacement of 0x00.
    */
   t_o = 0x48;
-  snprintf((char *)b + t_o, eos - t_o, "\x04\x03\x02\x01\xff" MSG_F, f->dir);
+  eos = (0x7ded-0x7c00); // 0x7ded is the maximum limit of string
+  snprintf((char *)b + t_o, eos - t_o, "\x04\x03\x02\x01\xff"
+    "\r\n"
+    "Sorry, could not load an operating system from\r\n"
+    "%s\r\n"
+    "Please check the IO.SYS file for corruption or replace with FreeDOS\r\n"
```
Comment (stsp) [inline src/base/misc/fatfs.c]: > or replace with FreeDOS

Cool! :)
OK, I'll apply that pearl. :)
But next time please suggest replacing with fdpp!

**Comment by stsp** [comment]
Thanks!

**Comment by andrewbird** [comment]
> But looking into this now, I
see you saying "dosemu boot block specific code",
which it is actually not: it was a hack you added
to pass NDD test. 

To clarify, I meant that it was code that was specific to the creation of the dosemu boot block and never used on the boot block derived from the 'boot.blk' file,  not that it created boot code.

**Comment by andrewbird** [comment]
No I wasn't saying that it changed anything, only that I think you read my comment as saying the NDD hack created DOS boot code, which is not what I meant.

**Comment by stsp** [comment]
I was trying to find out if you have something
against the aforementioned change. So seems not. :)

**Comment by andrewbird** [comment]
Ahh only that I'd like to move it back to group similar code together, but it's agreed there's no functional change.

**Comment by ecm-pushbx** [comment]
> > For example, if it was called from segment 0x70, then exit?
> 
> I think I observed that the early loader relocates itself to the top of low memory first. I can check tomorrow.

The loader indeed relocates itself to ((segment behind low memory end) & 0F000h) - 1000h, ie onto the last 64 KiB boundary that has a full 64 KiB of memory available.

**Pull request merged**